### PR TITLE
SDPAP-5913: Addition business fields to CMS user account.

### PR DESCRIPTION
**Pull request by henrytranvan**

### CODE REVIEW

File: config/install/user.role.site_admin.yml [hunk 1]
```diff
@@ -138,20 +410,65 @@ permissions:
...
   - 'view the administration theme'
+  - 'view tide_site preview links'
   - 'view timeline revisions'
+  - 'view translated_page revisions'
+  - 'view translated_vaccine_page revisions'
+  - 'view unpublished share link token entities'
+  - 'view vada_profile revisions'
+  - 'view vada_profile revisions'
+  - 'view vdrp_profile revisions'
+  - 'view vgso_profile revisions'
```
Comment (MdNadimHossain) [inline config/install/user.role.site_admin.yml:474]: please add the permissions that are needed. I can see there are lots of permissions which are only content-vic specific. Just add the permissions needed related to this PR please

File: tide_core.install [hunk 2]
```diff
@@ -1548,3 +1548,29 @@ function tide_core_update_8064() {
...
+ * Update user form display and authenticated user permission.
+ */
+function tide_core_update_8066() {
+  module_load_include('inc', 'tide_core', 'includes/helpers');
+  $config_location = [drupal_get_path('module', 'tide_core') . '/config/install'];
+  _tide_import_single_config('user.role.authenticated', $config_location);
+  _tide_import_single_config('user.role.site_admin', $config_location);
+  _tide_import_single_config('core.entity_form_display.user.user.default', $config_location);
+}
+
```
Comment (MdNadimHossain) [inline tide_core.install:1576]: please use `\Drupal::entityTypeManager()` to create the storage and field for both of the update hook. there are other examples in this file that you can follow. e.g - https://github.com/dpc-sdp/tide_core/blob/1082c243f94fce61cc242110595db4735169e3f5/tide_core.install#L1397

File: tide_core.install [hunk 3]
```diff
@@ -1548,3 +1548,29 @@ function tide_core_update_8064() {
     }
   }
 }
+
+/**
+ * Add contact field to user account(business name, phone, notes).
+ */
+function tide_core_update_8065() {
+  module_load_include('inc', 'tide_core', 'includes/helpers');
+  $config_location = [drupal_get_path('module', 'tide_core') . '/config/install'];
```
Comment (MdNadimHossain) [inline tide_core.install:1557]: this is deprecated from drupal/core 9.3 - `drupal_get_path('module',`
Please use the following one as from the next release we are updating every project to 9.4 - 
`$config_location = [\Drupal::service('extension.list.module')->getPath`
See this - https://www.drupal.org/node/2940438

**Review by MdNadimHossain — CHANGES_REQUESTED** [review]
Hi @henrytranvan 
There are few things that needs to be cleaned up. Please have a look and let me know if you have any question. Also as this is a core change and also in the next release it will go with drupal/core 9.4 update, can you please create a test PR in the content-reference site. There is a 9.4 test PR in the reference sit, you can branch out from that and use alias to apply this tide_core PR changes. Let me know if you need more info on that.  This is the reference site 9.4 branch that you can use - https://github.com/dpc-sdp/content-reference-sdp-vic-gov-au/pull/134

File: tide_core.install [hunk 4]
```diff
@@ -1548,3 +1548,36 @@ function tide_core_update_8064() {
...
+/**
+ * Add contact field to user account(business name, phone, notes).
+ */
+function tide_core_update_8065() {
+  module_load_include('inc', 'tide_core', 'includes/helpers');
+  $config_location = [\Drupal::service('extension.list.module')->getPath('tide_core') . '/config/install'];
+  $configs = [
+    'field.storage.user.field_business_contact_number' => 'field_storage_config',
+    'field.storage.user.field_business_name' => 'field_storage_config',
+    'field.storage.user.field_notes' => 'field_storage_config',
```
Comment (MdNadimHossain) [inline tide_core.install:1561]: So you are creating the field storage here, I don't see anything to create the fields though. e.g - `field.field.user.user.field_business_contact_number.yml`
`field.field.user.user.field_business_name.yml`
`field.storage.user.field_business_contact_number.yml`

File: tide_core.install [hunk 5]
```diff
@@ -1548,3 +1548,36 @@ function tide_core_update_8064() {
...
+/**
+ * Update user form display and authenticated user permission.
+ */
+function tide_core_update_8066() {
+  module_load_include('inc', 'tide_core', 'includes/helpers');
+  $config_location = [\Drupal::service('extension.list.module')->getPath('tide_core') . '/config/install'];
+  _tide_import_single_config('user.role.authenticated', $config_location);
+  _tide_import_single_config('user.role.site_admin', $config_location);
+  _tide_import_single_config('core.entity_form_display.user.user.default', $config_location);
+}
```
Comment (MdNadimHossain) [inline tide_core.install:1583]: Please do not use this `_tide_import_single_config` anymore, I will add a tag in this function to say deprecated
Also you can create the `entity_form_display` and `entity view display` using the `\Drupal::entityTypeManager()` Please check the following example - https://github.com/dpc-sdp/tide_landing_page/blob/develop/tide_landing_page.install#L1328

File: tide_core.install [hunk 6]
```diff
@@ -1548,3 +1548,36 @@ function tide_core_update_8064() {
...
+  }
+}
+
+/**
+ * Update user form display and authenticated user permission.
+ */
+function tide_core_update_8066() {
+  module_load_include('inc', 'tide_core', 'includes/helpers');
+  $config_location = [\Drupal::service('extension.list.module')->getPath('tide_core') . '/config/install'];
+  _tide_import_single_config('user.role.authenticated', $config_location);
```
Comment (MdNadimHossain) [inline tide_core.install:1580]: I do not think it needs permission for the authenticated user.

File: tide_core.install [hunk 7]
```diff
@@ -1548,3 +1548,36 @@ function tide_core_update_8064() {
...
+  _tide_import_single_config('user.role.site_admin', $config_location);
```
Comment (MdNadimHossain) [inline tide_core.install:1581]: does the site_admin needs all these permissions though ! isn't it like a normal other fields that just shows up and users can input things. I don't think it needs the field level permission. Can you please check if it works without these field level permission and user can input the values.

File: config/install/user.role.site_admin.yml [hunk 8]
```diff
@@ -142,6 +148,9 @@ permissions:
   - 'view all revisions'
   - 'view any unpublished content'
   - 'view any webform submission'
+  - 'view field_business_contact_number'
+  - 'view field_business_name'
+  - 'view field_notes'
   - 'view image_gallery revisions'
   - 'view landing_page revisions'
   - 'view landing_page revisions'
```
Comment (MdNadimHossain) [inline config/install/user.role.site_admin.yml:156]: don't think these permission are needed, please check the comment below.

File: config/install/user.role.authenticated.yml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+is_admin: false
+permissions:
+  - 'access content'
+  - 'access jsonapi resource list'
+  - 'edit own field_business_contact_number'
+  - 'edit own field_business_name'
+  - 'view media'
+  - 'view own field_business_contact_number'
+  - 'view own field_business_name'
+  - 'view published share link token entities'
```
Comment (MdNadimHossain) [inline config/install/user.role.authenticated.yml:16]: this one shouldn't require for authenticated users

File: composer.json [hunk 10]
```diff
@@ -79,6 +79,7 @@
         "drupal/jquery_ui_touch_punch": "^1.0",
         "drupal/admin_audit_trail": "^1.0",
         "drupal/block_inactive_users": "^2.0",
+        "drupal/field_permissions": "^1.2",
```
Comment (MdNadimHossain) [inline composer.json:82]: this is not needed, it i just normal new fields that getting added in the user form, please check if it works fine without this field level permissions.

**Review by MdNadimHossain — CHANGES_REQUESTED** [review]
Hi @henrytranvan 
Sorry but few more changes required. Please have a look at the comments and let me know if something does not make sense. Also it will be good to add a behat test so that it does a quick check on the fields in the user form if possible

**Review by MdNadimHossain — CHANGES_REQUESTED** [review]
Hey @henrytranvan 
looks good now. But can you please add a behat test to cover different permission scenario.

**Comment by vincent-gao** [comment]
hey @henrytranvan , could you please add details to the ticket.

**Review by MdNadimHossain — APPROVED** [review]
Looks good to me now, but please do add the details of this PR at the top as Vincent mentioned. e.g related Jira ticket, what changes you are making etc

**Pull request merged**